src/osd/PG.cc

```diff
@@ -1616,21 +1632,30 @@ void PG::activate(ObjectStore::Transaction& t,
       get_osdmap()->get_epoch(),
       activation_epoch));
   
-  // initialize snap_trimq
   if (is_primary()) {
-    dout(20) << "activate - purged_snaps " << info.purged_snaps
-	     << " cached_removed_snaps " << pool.cached_removed_snaps << dendl;
-    snap_trimq = pool.cached_removed_snaps;
-    interval_set<snapid_t> intersection;
-    intersection.intersection_of(snap_trimq, info.purged_snaps);
-    if (intersection == info.purged_snaps) {
-      snap_trimq.subtract(info.purged_snaps);
+    // initialize snap_trimq
+    if (get_osdmap()->require_osd_release < CEPH_RELEASE_MIMIC) {
+      dout(20) << "activate - purged_snaps " << info.purged_snaps
+	       << " cached_removed_snaps " << pool.cached_removed_snaps
+	       << dendl;
+      snap_trimq = pool.cached_removed_snaps;
     } else {
-      dout(0) << "warning: info.purged_snaps (" << info.purged_snaps
-	      << ") is not a subset of removed_snaps" << dendl;
-      snap_trimq.subtract(intersection);
-      assert(!cct->_conf->osd_debug_verify_cached_snaps);
+      auto& removed_snaps_queue = get_osdmap()->get_removed_snaps_queue();
+      auto p = removed_snaps_queue.find(info.pgid.pgid.pool());
+      snap_trimq.clear();
+      if (p != removed_snaps_queue.end()) {
+	dout(20) << "activate - purged_snaps " << info.purged_snaps
+		 << " removed_snaps " << p->second
+		 << dendl;
+	for (auto q : p->second) {
+	  snap_trimq.insert(q.first, q.second);
+	}
+      }
     }
+    interval_set<snapid_t> intersection;
+    intersection.intersection_of(snap_trimq, info.purged_snaps);
+    snap_trimq.subtract(intersection);
+    info.purged_snaps.swap(intersection);
```

Comment:
I’m not getting how we can do this swap on pre-mimic?

Reply:
this is actually identical behavior to the old code, except the old code had different debug output if the old purged_snaps wasn't a subset of removed_snaps (if they are == then substracting intersection is the same thing):
<pre>
  // initialize snap_trimq
   if (is_primary()) {
-    dout(20) << "activate - purged_snaps " << info.purged_snaps
-	     << " cached_removed_snaps " << pool.cached_removed_snaps << dendl;
-    snap_trimq = pool.cached_removed_snaps;
-    interval_set<snapid_t> intersection;
-    intersection.intersection_of(snap_trimq, info.purged_snaps);
-    if (intersection == info.purged_snaps) {
-      snap_trimq.subtract(info.purged_snaps);
     } else {
-      dout(0) << "warning: info.purged_snaps (" << info.purged_snaps
-	      << ") is not a subset of removed_snaps" << dendl;
-      snap_trimq.subtract(intersection);
-      assert(!cct->_conf->osd_debug_verify_cached_snaps);
     }
 }
</pre>

Reply:
I never liked this code (it's too many swaps), but I'm not seeing how we want to swap the new intersection set with the info's purged_snaps. I don't *think* we did that before?

Reply:
It's because removed_snaps_queue will shrink as the purged snaps get pruned from the osdmap.  when that happens our purged_snaps here also needs to shrink (so we don't keep reporting old news).

Reply:
I am good with us getting the intersection and then trimming the snap_trimq by that value.

But the part where we swap our intersection into the purged_snaps is confusing me. On Mimic code I think it makes sense, because we have set up the snap_trimq from what the monitor knows considers removed_snaps (so we can dump purged_snaps the monitor agrees are purged). But for *pre*-Mimic, I think we're just tossing out information about which snapshots still exist or not.

Reply:
Yeah, that makes sense. made that condition on >= mimic. It should never shrink pre-mimic.